examples/chrono/chrono.cpp

```diff
@@ -0,0 +1,219 @@
+  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) | mcal::reg::bval5);
+
+  for(;;)
+  {
+    // Toggle the LED via direct memory access.
+    *reinterpret_cast<volatile port_type*>(mcal::reg::portb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) ^ mcal::reg::bval5);
+
+    local::delay_1s();
+  }
+}
+
+auto local::delay_1s() noexcept -> void
+{
+  // Obtain an initial value of std::chrono::high_resolution_clock::now().
+  const auto start = clock_type::now();
+
+  bool is_timeout = false;
+
+  // Wait for 1s, while triggering the watchdog timer in the loop.
+  while(!is_timeout)
+  {
+    mcal::wdg::trigger();
+
+    const auto delta = std::chrono::duration_cast<std::chrono::seconds>(clock_type::now() - start);
+
+    is_timeout = (delta != static_cast<std::chrono::seconds>(0U));
+  }
+}
+
+extern "C"
+void __vector_16(void) ATTRIBUTE(signal, used, externally_visible);
+
+void __vector_16(void)
+{
+  // Increment the 64-bit system tick with 0x80, representing 128 microseconds.
+  const auto new_tick = static_cast<mcal::gpt::value_type>(local::mcal_gpt_system_tick + UINT8_C(0x80));
+
+  local::mcal_gpt_system_tick = new_tick;
+}
+
+auto mcal::gpt::init() -> void
+{
+  // Clear the timer0 overflow flag.
+  *reinterpret_cast<volatile std::uint8_t*>(mcal::reg::tifr0) = UINT8_C(0x01);
+
+  // Enable the timer0 overflow interrupt.
+  *reinterpret_cast<volatile std::uint8_t*>(mcal::reg::timsk0) = UINT8_C(0x01);
+
+  // Set the timer0 clock source to f_osc/8 = 2MHz and begin counting.
+  *reinterpret_cast<volatile std::uint8_t*>(mcal::reg::tccr0b) = UINT8_C(0x02);
+}
+
+auto mcal::gpt::get_time_elapsed() -> mcal::gpt::value_type
+{
+  // Return the system tick using a multiple read to ensure data consistency.
+
+  using timer_register_type = std::uint8_t;
+
+  // Do the first read of the timer0 counter and the system tick.
+  const timer_register_type tim0_cnt_1 = *reinterpret_cast<volatile timer_register_type*>(mcal::reg::tcnt0);
+  const auto                sys_tick_1 = local::mcal_gpt_system_tick;
+
+  // Do the second read of the timer0 counter.
+  const timer_register_type tim0_cnt_2 = *reinterpret_cast<volatile timer_register_type*>(mcal::reg::tcnt0);
+
+  // Perform the consistency check.
+  const mcal::gpt::value_type consistent_microsecond_tick =
+    ((tim0_cnt_2 >= tim0_cnt_1) ? static_cast<mcal::gpt::value_type>(sys_tick_1                  | static_cast<std::uint8_t>(tim0_cnt_1 >> 1U))
+                                : static_cast<mcal::gpt::value_type>(local::mcal_gpt_system_tick | static_cast<std::uint8_t>(tim0_cnt_2 >> 1U)));
+
+  return consistent_microsecond_tick;
+}
+
+auto mcal::wdg::init() -> void
+{
+  // Read the MCU status register.
+  volatile const std::uint8_t mcu_status_register = *reinterpret_cast<volatile std::uint8_t*>(mcal::reg::mcusr);
+
+  static_cast<void>(mcu_status_register);
```

Comment:
Why is the status register read here? The result is not used. Does the read have any hardware side effects?

Reply:
> Why is the status register read here? The result is not used. Does the read have any hardware side effects?

I forgot and I will check. Looking at my records, I first wrote a version of that code around 2005 in assembly. I'll see if it is needed in newer manuals.

Reply:
>> Why is the status register read here? 

> I forgot and I will check. Looking at my records, I first wrote a version of that code around 2005 in assembly. I'll see if it is needed in newer manuals.

Ah... Evidently readint the status register was a placeholder for myself in case I had ever wanted to query the reset reason.

Looking at Sect. 15.8.2 in Microchip's "Atmel-2545W-ATmega48/V / 88/V / 168/V_Datasheet_Complete-11/2016", there seems to be no need to read the MCU status register.

I think, however, it's a good idea to clear the WDRF flag and will do so.

Reply:
And finally for the elephant in the room --- why use a watchdog timer at all in a simple example meant to exemplify aspects of using `<chrono>` ?

The possible choices are: Include watchdog? yes/no.

This is optional. When using `<chrono>` an example makes sense it there is a little microcosm of a timer structure running. Whenever timers are involved (and using them for delays), it makes sense to trigger a watchdog.

So I opted for use watchdog: Yes.